packages/block-editor/src/components/block-toolbar-link-control/link-input-toolbar-item.js

```diff
@@ -0,0 +1,153 @@
+
+/**
+ * WordPress dependencies
+ */
+import {
+	__experimentalToolbarItem as ToolbarItem,
+	Popover,
+	__experimentalToolbarContext as ToolbarContext,
+	__experimentalInputControl as InputControl,
+	Icon,
+	Spinner,
+} from '@wordpress/components';
+import { link as linkIcon } from '@wordpress/icons';
+import { useContext, useEffect, useState, useRef } from '@wordpress/element';
+import { useDispatch } from '@wordpress/data';
+
+/**
+ * Internal dependencies
+ */
+import LinkControlSearchResults from '../link-control/search-results';
+import LinkControlSearchInput from '../link-control/search-input';
+import useCreatePage from '../link-control/use-create-page';
+import ToolbarLinkControlContext from './context';
+
+export default function LinkInputToolbarItem() {
+	const toolbar = useContext( ToolbarContext );
+
+	const [ editUrl, setEditUrl ] = useState( '' );
+	return (
+		<ToolbarItem>
+			{ ( htmlProps ) => (
+				<div
+					{ ...htmlProps }
+					className="toolbar-link-control__input-wrapper"
+				>
+					<ToolbarWidget
+						as={ ToolbarLinkEditorControl }
+						{ ...toolbar }
+						value={ editUrl }
+						onChange={ setEditUrl }
+					/>
+				</div>
+			) }
+		</ToolbarItem>
+	);
+}
+
+const ToolbarLinkEditorControl = function ( props ) {
+	const {
+		createSuggestion,
+		currentLink,
+		updateCurrentLink,
+		preferredDropdown,
+		setPreferredDropdown,
+	} = useContext( ToolbarLinkControlContext );
+
+	const { createPage, isCreatingPage, errorMessage } = useCreatePage(
+		createSuggestion
+	);
+
+	const { createErrorNotice } = useDispatch( 'core/notices' );
+	useEffect( () => {
+		if ( errorMessage ) {
+			createErrorNotice( errorMessage, { type: 'snackbar' } );
+		}
+	}, [ errorMessage ] );
+
+	const searchInputRef = useRef();
+
+	return (
+		<LinkControlSearchInput
+			ref={ searchInputRef }
+			currentLink={ currentLink }
+			placeholder="Start typing"
+			renderSuggestions={ renderSuggestions }
+			value={ currentLink.url }
+			onCreateSuggestion={ createPage }
+			onChange={ ( url ) => {
+				setPreferredDropdown( 'suggestions' );
+				updateCurrentLink( { url } );
+			} }
+			onSelect={ ( link ) => {
+				updateCurrentLink( link, true );
+			} }
+			showInitialSuggestions={ false }
+			allowDirectEntry
+			showSuggestions={ preferredDropdown === 'suggestions' }
+			withCreateSuggestion
+			renderControl={ ( controlProps, inputProps, isLoading ) => {
+				return (
+					<InputControl
+						{ ...controlProps }
+						{ ...inputProps }
+						{ ...props }
+						className="toolbar-link-control__input-control"
+						value={ inputProps.value }
+						onKeyDown={ ( event ) => {
+							inputProps.onKeyDown( event );
+							// LinkControlSearchInput renders a form which is normally submitted with an Enter key.
+							// In this context however, Reakit calls .preventDefault() on the enter keydown event
+							// so we need to select focused suggestion manually.
```

Comment:
With the combobox pattern it is expected that pressing enter on a focused result will select that result. Removing this behaviour effectively breaks the accessibility of this component. Do we really need Reakit here? I'd expect the arrow keyboard navigation between controls to be dealt with by the parent toolbar. If not, we also have our own implementation of `RovingTabIndex` in the [TreeGrid](https://github.com/WordPress/gutenberg/blob/d15a208001f1d728d4f7ffb6b61239fc3fad5fbc/packages/components/src/tree-grid/README.md#L129) component.

Reply:
I'm not sure I follow, I may have confused you with that comment. The current behavior is that typing in something in that input triggers a dropdown/combobox which is keyboard-navigable. You may use arrow keys to change the focused result and then use the enter key to select the currently focused one, which sounds in line with what you described, see how the code reads without all the comments:

```js
if ( event.key === 'Enter' ) {
	setTimeout( () => {
		if ( searchInputRef?.current ) {
			searchInputRef.current.selectFocusedSuggestion();
		}
	} );
}
```

> Do we really need Reakit here?

What I meant was that in this use-case, the `<LinkControlSearchInput />` is embedded within an accessible Reakit toolbar. Both `LinkControlSearchInput` and the toolbar have event handlers for the same events: keyDown, change, focus. Normally in this context Reakit calls `.preventDefault()` whenever it encounters a keyDown event:
https://github.com/reakit/reakit/blob/02d13512c49f44305ecfda8bcd68531cd45cb33b/packages/reakit/src/Composite/CompositeItemWidget.ts#L82